Implemented mock data for Feed component using faker.js

alextrieu opened the pull request:
This pull request integrates the faker.js library into our Feed component to generate mock data. 

The changes introduced in this PR are as follows:

- Implemented a function to generate mock data for each post
- Used the generated mock data in the Feed component
- Passed the mock data as props to child components for display

alextrieu committed:
Implemented mock data for Feed component using faker.js

ruzicic commented on src/App.tsx:
```diff
@@ -1,13 +1,27 @@
+function usePosts(count: number): Post[] {
+  const [posts, setPosts] = useState<Post[]>([]);
+
+  useEffect(() => {
```
Since all of this happens synchronously, you can remove the `useEffect` (and useState) completely and have this hook immediately return the mocked post data.

ruzicic commented on src/App.tsx:
```diff
@@ -1,13 +1,27 @@
     <main className="app-container">
       <NavigationBar />
-      <Feed />
+      <Feed postsData={posts} />
```
[nitpicking] Either just `posts` or `data` is sufficient here. Keep the naming simple 💅

ruzicic commented on src/data/MockData.tsx:
```diff
@@ -0,0 +1,44 @@
+  profilePic: string;
+};
+
+export type Post = {
```
[nitpicking] I like reusing types like this one, but this probably isn't the best place for the declaration of these types.

Maybe a common directory with types and `post.types.ts` would make more sense.

ruzicic commented on src/data/MockData.tsx:
```diff
@@ -0,0 +1,44 @@
+import { faker } from "@faker-js/faker";
+
+type Comment = {
```
Instead of redeclaring same attributes, consider using [Pick](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys) where it makes sense (like here)

alextrieu replied:
Would that mean the commentText property from the type Comment be moved over to type User? 

Then using the Pick type to pick the ‘username’ and ‘commentText’ properties over to type Comment? 

type Comment = Pick<User, "username" | "commentText">;

export type User = {
  username: string;
  profilePic: string;
  commentText?: string;
};

alextrieu replied:
Or would this do?

type Comment = Pick<User, "username"> & { commentText: string };

ruzicic replied:
This question has more to do with business logic than typescript :)

I'd say the second suggestion

```ts
type Comment = Pick<User, "username"> & { text: string };
```

ruzicic commented on src/data/MockData.tsx:
```diff
@@ -0,0 +1,44 @@
+  timestamp: Date;
+};
+
+const generatePostData = (): Post => {
```
Great job!

ruzicic commented on src/components/Feed/PostContainer/PostCaption.tsx:
```diff
@@ -1,17 +1,31 @@
 import styles from "./PostContainer.module.css";
 
-const PostCaption: React.FC = () => {
+type PostCaptionProps = {
```
This may be a good place to use `Pick` from `Post`

ruzicic commented on src/components/Feed/PostContainer/PostCaption.tsx:
```diff
@@ -1,17 +1,31 @@
+  const maxLength = 80;
+
+  function handleExpandClick() {
+    setExpanded(!isExpanded);
```
If you want to toggle the previous value, this is not the best way to do it.

There are two alternatives:

1) use previous value `setExpanded(prev => !prev)`
2) wrap your handleExpandClick with `useCallback` and add `isExpanded` as a dependency

ruzicic commented on src/components/Feed/PostContainer/PostCaption.tsx:
```diff
@@ -1,17 +1,31 @@
       <a href="#" className={styles.profileLinkButton}>
         {" "}
-        <span style={{ fontWeight: "bold" }}>usaswimming </span>
+        <span style={{ fontWeight: "bold" }}>{username} </span>
```
Instead of <span /> + inline bold, consider using `<strong>` or `<em>`

ruzicic commented on src/components/Feed/PostContainer/PostLikes.tsx:
```diff
@@ -1,23 +1,25 @@
-        </button>
+        {profilePics.map((pic, index) => (
+          <button key={index} className={styles.profilePicButton}>
+            <img src={pic.profilePic}></img>
```
`<img />` is a self-closing tag

alextrieu committed:
Refactored usePosts hook for synchronous data generation

alextrieu committed:
Refactor prop name from dataPosts to data in Feed component

alextrieu committed:
Change image tag to self-closing

alextrieu committed:
Refactor span to strong tag for username emphasis

alextrieu committed:
Move type declarations to separate file

Created a new folder 'types' and moved all type declarations to 'PostTypes.tsx'

alextrieu committed:
Correct import path for Post in App.tsx

alextrieu committed:
Update PostCaptionProps type definition using the Pick utility type

alextrieu committed:
Implement dynamic username display for post likes

alextrieu committed:
Use previous state to toggle isExpanded in PostCaption

alextrieu committed:
Update props in PostLikes component
